### PR TITLE
ARROW-902: [C++] Script for downloading all thirdparty build dependencies and configuration for offline builds

### PR DESCRIPTION
**Pull request by wesm**
This patch provides a script for downloading all the third party dependencies to a particular location, and new environment variables that can be used to direct the build system to use local files for ExternalProject builds instead of accessing the internet.

### CODE REVIEW

File: cpp/thirdparty/download_dependencies.sh [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+SNAPPY_VERSION=1.1.3
+BROTLI_VERSION=v0.6.0
+LZ4_VERSION=1.7.5
+ZLIB_VERSION=1.2.8
+ZSTD_VERSION=1.2.0
+PROTOBUF_VERSION=2.6.0
+GRPC_VERSION=1.12.1
+ORC_VERSION=1.5.1
+
+BOOST_UNDERSCORE_VERSION=`echo $BOOST_VERSION | sed 's/\./_/g'`
```
Comment (xhochy) [inline cpp/thirdparty/download_dependencies.sh]: A nice improvement for this script would be to only download files that were not yet downloaded. Either by checking for their existence or by using `wget -c`

File: cpp/thirdparty/download_dependencies.sh [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+SOURCE_DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
+
+if [ "$#" -ne 1 ]; then
+  echo "Usage: $0 <destination-directory>"
+  exit
+fi
+
+_DST=$1
+
+BOOST_VERSION=1.67.0
```
Comment (pitrou) [inline cpp/thirdparty/download_dependencies.sh]: Is there a way we can avoid duplicating version numbers and download URLs between this and the CMake files?

Comment (wesm) [reply cpp/thirdparty/download_dependencies.sh]: I don't care for the idea of maintaining the version numbers in two places either. I'm not sure how to get data from a file _in_ to CMake. I have done a bit of Google searching and I can't really find anything conclusive. CMake does have rudimentary string / regular expression parsing, that's probably the only way. I will take a crack at it and put a file like `thirdparty/versions.txt`

**Comment by wesm** [comment]
OK, I think I've addressed all the comments so far. I will improve the documentation / README and then this should be good to go

**Comment by wesm** [comment]
+1

**Pull request closed without merging**